### PR TITLE
lib/ramfs: Fix UAF bug when deleting open file

### PR DESCRIPTION
**Pull request by andreittr**
### Description of changes

Previously ramfs would eagerly free filesystem nodes when deleting a file regardless whether the file is open, leading to a possible use-after-free bug if the open file is later used. This change fixes this issue by marking nodes as deleted after they have been removed from the filesystem, then only releasing resources lazily once all open references have been garbage-collected.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A

### Additional configuration

Test with:
```c
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
#define FN "tmpfile"

	int f;
	int r;
	void *buf;

	if ((f = open(FN, O_RDWR|O_CREAT, 0777)) < 0) {
		puts("Badf");
		return 1;
	}

	if ((r = unlink(FN))) {
		printf("Bad unlink %d\n", r);
		return 1;
	}

	buf = malloc(4000);
	printf("%p\n", buf);
	memset(buf, 0xff, 4000);

	if ((r = write(f, "asdf", 4)) < 0) {
		printf("Bad write %d\n", r);
		return 1;
	}

	if ((r = ftruncate(f, 0))) {
		printf("Bad truncate %d\n", r);
		return 1;
	}

	if ((r = close(f))) {
		printf("Bad close %d\n", r);
		return 1;
	}

	return 0;
}
```

### CODE REVIEW

File: lib/ramfs/ramfs_vnops.c [hunk 1]
```diff
@@ -55,6 +55,12 @@
 #include <fcntl.h>
 #include <vfscore/fs.h>
 
+/* 16 bits are enough for file mode as defined by POSIX, rest we can use */
```
Comment (RaduNichita) [inline lib/ramfs/ramfs_vnops.c]: It not very clear to me at what  "rest we can use". Could you be more explicit there, please?

**Review by RaduNichita — COMMENTED** [review]
It works for me, I will approve after discussing my only comment.

File: lib/ramfs/ramfs_vnops.c (hunk 1 above)
Comment (andreittr) [reply lib/ramfs/ramfs_vnops.c]: I'll add a "the" for grammar pedantry.

**Review by DeliaPavel — APPROVED** [review]
Looks good to me!

Reviewed-by: Delia Pavel <delia_maria.pavel@stud.acs.upb.ro>

**Review by RaduNichita — APPROVED** [review]
All good also from my side.

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request closed without merging**